In [1]:
from pyspark.sql import SparkSession
from pyspark.sql import functions as F
import pandas as pd
import matplotlib.pyplot as plt
import os

In [2]:
# Initialize Spark session
spark = SparkSession \
    .builder \
    .config("spark.executor.memory", "8g") \
    .config("spark.driver.memory", "16g") \
    .config("spark.memory.fraction", "0.8") \
    .getOrCreate()

/opt/conda/lib/python3.7/site-packages/pyspark/bin/load-spark-env.sh: line 68: ps: command not found
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


25/12/02 23:53:03 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [3]:
os.listdir('/standard/ds7200-apt4c/isaac_yuyang_final_project/')

['flights_2022.parquet',
 'flights_2021.parquet',
 'flights_2020.parquet',
 'flights_2018.parquet',
 'flights_2019.parquet',
 'Airlines.csv']

In [4]:
airlines = spark.read.load('/standard/ds7200-apt4c/isaac_yuyang_final_project/Airlines.csv', header=True, format='csv', sep=',')

In [5]:
airlines.show(10)

+----+--------------------+
|Code|         Description|
+----+--------------------+
| 02Q|       Titan Airways|
| 04Q|  Tradewind Aviation|
| 05Q| Comlux Aviation, AG|
| 06Q|Master Top Linhas...|
| 07Q| Flair Airlines Ltd.|
| 09Q|      Swift Air, LLC|
| 0BQ|                 DCA|
| 0CQ|ACM AIR CHARTER GmbH|
| 0GQ|Inter Island Airw...|
| 0HQ|Polar Airlines de...|
+----+--------------------+
only showing top 10 rows



In [7]:
airlines.count()

1571

In [3]:
# Read in data
flights_2018 = spark.read.parquet('/standard/ds7200-apt4c/isaac_yuyang_final_project/flights_2018.parquet')
flights_2019 = spark.read.parquet('/standard/ds7200-apt4c/isaac_yuyang_final_project/flights_2019.parquet')
flights_2020 = spark.read.parquet('/standard/ds7200-apt4c/isaac_yuyang_final_project/flights_2020.parquet')
flights_2021 = spark.read.parquet('/standard/ds7200-apt4c/isaac_yuyang_final_project/flights_2021.parquet')
flights_2022 = spark.read.parquet('/standard/ds7200-apt4c/isaac_yuyang_final_project/flights_2022.parquet')

In [9]:
flights_2018.columns

['FlightDate',
 'Airline',
 'Origin',
 'Dest',
 'Cancelled',
 'Diverted',
 'CRSDepTime',
 'DepTime',
 'DepDelayMinutes',
 'DepDelay',
 'ArrTime',
 'ArrDelayMinutes',
 'AirTime',
 'CRSElapsedTime',
 'ActualElapsedTime',
 'Distance',
 'Year',
 'Quarter',
 'Month',
 'DayofMonth',
 'DayOfWeek',
 'Marketing_Airline_Network',
 'Operated_or_Branded_Code_Share_Partners',
 'DOT_ID_Marketing_Airline',
 'IATA_Code_Marketing_Airline',
 'Flight_Number_Marketing_Airline',
 'Operating_Airline',
 'DOT_ID_Operating_Airline',
 'IATA_Code_Operating_Airline',
 'Tail_Number',
 'Flight_Number_Operating_Airline',
 'OriginAirportID',
 'OriginAirportSeqID',
 'OriginCityMarketID',
 'OriginCityName',
 'OriginState',
 'OriginStateFips',
 'OriginStateName',
 'OriginWac',
 'DestAirportID',
 'DestAirportSeqID',
 'DestCityMarketID',
 'DestCityName',
 'DestState',
 'DestStateFips',
 'DestStateName',
 'DestWac',
 'DepDel15',
 'DepartureDelayGroups',
 'DepTimeBlk',
 'TaxiOut',
 'WheelsOff',
 'WheelsOn',
 'TaxiIn',
 'CRS

In [30]:
flights_2018.groupBy('OriginCityName').count().show()

+------------------+-----+
|    OriginCityName|count|
+------------------+-----+
|   Gainesville, FL| 2092|
|      Richmond, VA|18967|
|       Ontario, CA|18590|
|        Tucson, AZ|13103|
|     Pago Pago, TT|  122|
|  Myrtle Beach, SC| 8961|
|       Medford, OR| 6696|
|  Palm Springs, CA| 8041|
|       Redding, CA| 1169|
|     St. Cloud, MN|  145|
|       Durango, CO| 2851|
|   Devils Lake, ND|  626|
|      Gustavus, AK|   31|
|        Mobile, AL| 4027|
|Corpus Christi, TX| 4291|
|       Dubuque, IA|  435|
|      Columbus, GA|  606|
|    Huntsville, AL| 6156|
|    Fort Myers, FL|25611|
|     Pensacola, FL| 7248|
+------------------+-----+
only showing top 20 rows



In [4]:
# Combine all years
combined = flights_2018.union(flights_2019).union(flights_2020).union(flights_2021).union(flights_2022)

In [14]:
combined.select('ArrDelay', 'ArrDelayMinutes').show(20)

+--------+---------------+
|ArrDelay|ArrDelayMinutes|
+--------+---------------+
|    -8.0|            0.0|
|    -6.0|            0.0|
|    -2.0|            0.0|
|   -11.0|            0.0|
|    -1.0|            0.0|
|    22.0|           22.0|
|    -1.0|            0.0|
|    -9.0|            0.0|
|    null|           null|
|    22.0|           22.0|
|   -12.0|            0.0|
|   -13.0|            0.0|
|   -10.0|            0.0|
|   -13.0|            0.0|
|    16.0|           16.0|
|   105.0|          105.0|
|   -11.0|            0.0|
|     8.0|            8.0|
|     1.0|            1.0|
|   -10.0|            0.0|
+--------+---------------+
only showing top 20 rows



In [16]:
numeric_cols = [field.name for field in combined.schema.fields 
                if field.dataType.typeName() in ['integer', 'double', 'float', 'long']]
correlations = []
for col in numeric_cols:
    if col not in ['ArrDelay', 'ArrDelayMinutes', 'ArrivalDelayGroups', 'ArrDel15']:
        corr = combined.stat.corr(col, 'ArrDelay')
        correlations.append((col, corr))
# Sort by absolute correlation value
correlations.sort(key=lambda x: abs(x[1]), reverse=True)

# Display results
for col, corr in correlations:
    print(f"{col}: {corr:.4f}")

DepDelay: 0.9572
DepDelayMinutes: 0.9540
DepartureDelayGroups: 0.8092
DepDel15: 0.5540
TaxiOut: 0.2005
DepTime: 0.1128
WheelsOff: 0.1092
TaxiIn: 0.1055
CRSDepTime: 0.0826
CRSArrTime: 0.0690
ActualElapsedTime: 0.0465
WheelsOn: 0.0350
DOT_ID_Marketing_Airline: 0.0340
ArrTime: 0.0287
DOT_ID_Operating_Airline: 0.0266
__index_level_0__: 0.0181
DestWac: -0.0128
AirTime: 0.0116
CRSElapsedTime: -0.0106
Flight_Number_Operating_Airline: 0.0089
Flight_Number_Marketing_Airline: 0.0089
OriginAirportSeqID: -0.0081
OriginAirportID: -0.0081
OriginCityMarketID: -0.0080
DistanceGroup: -0.0072
DestStateFips: 0.0071
Distance: -0.0071
OriginWac: -0.0071
DayOfWeek: 0.0048
Year: -0.0043
Quarter: 0.0042
Month: 0.0042
OriginStateFips: 0.0023
DivAirportLandings: -0.0022
DayofMonth: -0.0020
DestAirportSeqID: -0.0017
DestAirportID: -0.0017
DestCityMarketID: 0.0000


It looks like the only variables that are very strongly correlated with arrival delay are variables related to departure delay, which makes sense, but isn't particularly useful for predicting flight status, since if it's delayed departing then it isn't suprising that it is delayed arriving.

`TaxiOut`, which is the time in minutes that it took to taxi out from the departure gate, has a correlation of 0.2, which shows it could maybe be a useful predictor.

In [17]:
# Group by airline and date, calculate proportion of delayed flights
delay_prop = combined.groupBy('Airline') \
                .agg(F.mean('ArrDel15').alias('delay_proportion'))

In [19]:
delay_prop.show(30)

+--------------------+-------------------+
|             Airline|   delay_proportion|
+--------------------+-------------------+
|GoJet Airlines, L...|0.19908022176708454|
|   Endeavor Air Inc.|0.12985950829954618|
|       Allegiant Air|0.24518292119772414|
|SkyWest Airlines ...|0.16295041949138497|
|      Virgin America|  0.200548009094619|
|         Horizon Air|0.14094595742734087|
|United Air Lines ...|0.18529587619189977|
|Air Wisconsin Air...|0.18147947395508252|
|Trans States Airl...|0.25638202087071427|
|    Compass Airlines|0.18166881590919504|
|Peninsula Airways...| 0.3724378543392935|
|         Comair Inc.|0.17565905734565473|
|Frontier Airlines...|0.24292174526197638|
|            Cape Air|0.08942598187311178|
|Southwest Airline...|  0.181112993616926|
|ExpressJet Airlin...| 0.2155240719497629|
|     JetBlue Airways| 0.2560654332718499|
|Commutair Aka Cha...| 0.2628661437624129|
|Empire Airlines Inc.|0.15779774245565537|
|           Envoy Air| 0.1739135901212694|
|Capital Ca

By grouping by airline and taking the mean of the delayed arrivals, we can find the proportion of flights for each airline that are delayed by at least 15 minutes. Some outliers appear to be Cape Air with 8.9%, Endeavor Airlines with 12.9%, and Peninsula, with  37%. Most of the rest of the airlines are somewhere around 20%, $\pm$ ~6%.